Merging with distributed update PR

x2048 opened the pull request:
Hey Liso, the core developers would like to keep the PRs separate, so I'll need to revert the changes to distributed update PR :(

Please merge these changes into your branch to keep it up to date with minetest/minetest#11422.

x2048 committed:
Distribute shadow map update over multiple frames

Reduces stutter and freezes when playing.

x2048 committed:
Code cleanup according to review comments

Naming.
Loop optimization.
Use of STL.
Member initialization and access.

x2048 committed:
Avoid shadow flicker when camera offset changes

Adjust view matrix for the current shadow frustum so that previously generated SM
texture applies correctly to the view.

x2048 committed:
Split shadow map update by block position in light space.

Removes artifacts caused by incorrect Z-buffer scale between section updates.

x2048 committed:
Distribute SM update using texture blending

This reverts changes to shaders and splitting shadow frustum.
Light/shadow frustum updates are still made in transaction.

Wuzzy2 committed:
Add wallmounted support for plantlike and plantlike_rooted nodes (#11379)

x2048 committed:
Use single colored SM texture

There is no need in using two textures since we render colored shadows in one pass.

x2048 committed:
Merge performance and quality fixes from 0xLiso:PSM

Co-Authored-By: Liso <anlismon@gmail.com>

x2048 committed:
Calibrate and optimize soft shadows for the new light frustum

SmallJoker committed:
CSM: Do not index files within hidden directories

CSM would previously scan for files within .git or .svn directories, and also special files such as .gitignore

lhofhansl commented:
FWIW... This was just my suggestion. (Not sure how the other core devs feel.)

Edit: Maybe discuss on IRC...?
(Personally I'm a fan of multiple smaller PRs as it makes both review and testing easier. But if that leads to unreasonable burden or complexity larger PRs are certainly OK.)

x2048 commented:
Too bad I can't join on IRC, but I'm happy to bring a quick discussion in minetest/minetest#11422.

sfan5 committed:
Fix no locales being generated when APPLY_LOCALE_BLACKLIST=0

Also enable `ky` which appears to work fine.

hecktest committed:
Remove unused header includes

hecktest committed:
Remove unused ITextSceneNode header (#11476)

Co-authored-by: hecktest <>

random-geek committed:
Document glasslikeliquidlevel merge bits (#11479)

x2048 committed:
Add smooth light-shadow transition at noon (#11430)

Node faces with normals pointing East/West (+X/-X) will transition between light
and shadow at noon. This code makes the transition smooth.

x2048 committed:
Distribute shadow map update over multiple frames to reduce stutter (#11422)

Reduces stutter and freezes when playing.

 * Maintains double SM and SM Color textures
 * Light frustum update triggers incremental generation of shadow map into secondary 'future' textures.
 * Every incremental update renders a portion of the shadow draw list (split equally).
 * After defined number of frames (currently, 4), 'future' and 'current' textures are swapped, and DirectionalLight 'commits' the new frustum to use when rendering shadows on screen.

Co-authored-by: sfan5 <sfan5@live.de>

lhofhansl commented:
#11422 got merged. If you guys still have energy we can continue with this one now.

x2048 committed:
Merge changes from minetest:master
